### ISSUE TITLE
Doubledrop, Tripledrop & Superbreaker.

### ISSUE THREAD

**Issue by Rudomeister** [mcMMO-Dev/mcMMO#2544]
Just filing in about the doubledrop bug with new stonetypes. Random stonetype dropping out of Stone when getting doubledrop. At doubledrop stones are bouncing away, preferly into a lavapool nearby like a magnet effect. Same is about obsidian as well, it doubledrops into lava below, makes obsidian mining more or less useless around bedrock layer.

**Comment by astemus** [comment]
The bonus drops are spawning inside the block before it's removed from the world, so it's flung away as if it's suffocating.  It's rather annoying, and it's a bug that has been around for a long time now.

**Comment by nextgenerati0n** [comment]
I seem to have the same issue. I thought obsidian simply wasn't dropping however it could be the the above comment it what's really occurring. Could really do with this looking into as it's pretty much useless for people to try and mine obsidian with this happening.

**Comment by SidShakal** [comment]
I'm surprised this bug hasn't received more attention. When I first encountered it last year, I whipped up a workable patch on my own that satisfied my friends and me. I simply wrapped all the block-based input locations for calls to the the Misc.dropItem family of methods in a method that shifted the location passed to the dropItem method from the low corner of the block to the center of the block.

I believe what happens is the double-dropped items are dropped at the low corner of the block they were mined from. Drop combining often causes it to gobble up the drop that mining spawns normally. Then the combined drop, still at the low corner of the block and possibly colliding with other blocks that are not air blocks, gets the item-is-in-a-block collision treatment, and depending on which block it is considered inside, undesirable things are likely to happen:

- If you are mining upward in a cave, drops often end up in the next open space above, even though that could be dozens of meters above you. (Some item elevators exploit this same effect of the item-is-in-a-block collision algorithm to move items upward.)
- If you're mining obsidian near lava, oftentimes you're out of luck because the double-drop obsidian grabs the normally dropped obsidian and goes all murder-suicide by colliding with a lava block which is one of the eight blocks that meet at the corner of the obsidian block you just mined.
- Otherwise, you're likely to get drops being flung wildly one way or another to get them out of the space occupied by the block(s) they were spawned in.

I haven't walked through all the code to confirm that this is entirely accurate, and I'm not _intimately_ familiar with the item-is-in-a-block collision algorithm, but this model of what might cause the undesired behavior seems to explain it, and adjusting the item spawn location by shifting it from the low corner of the mined block to the center of that same block seems to get rid of the problems.

Ah. I just looked into it, and it appears that my solution of setting the drop position to the center of the block is a little too constant to pass as natural behavior.
To mimic the natural spawning of an item drop caused by the breaking of a block, the item drop needs to be spawned at a uniformly random location within a cube centered on the block's center, with sides that are half a block in length.

I'll get to work on this now and submit a pull request when it's done.
